packages/zoe/state.js

```diff
@@ -1,190 +1,259 @@
+    update: (handle, partialRecord) => {
+      const record = handleToRecord.get(handle);
+      const updatedRecord = {
+        ...record,
+        ...partialRecord,
+      };
+      table.validate(updatedRecord);
+      handleToRecord.set(handle, updatedRecord);
+      return record;
     },
+    // eslint-disable-next-line no-use-before-define
+  });
+
+  const customMethodsTable = harden({
+    ...makeCustomMethodsFn(table),
+    ...table,
   });
-  return {
-    adminState,
-    readOnlyState,
+  return customMethodsTable;
+};
+
+const validateProperties = (expectedProperties, obj) => {
+  // add handle to expected properties
+  expectedProperties.push('handle');
+  const actualProperties = Object.getOwnPropertyNames(obj);
+  insist(
+    actualProperties.length === expectedProperties.length,
+  )`the actual properties (${actualProperties}) did not match the expected properties (${expectedProperties})`;
+  for (const prop of actualProperties) {
+    insist(
+      expectedProperties.includes(prop),
+    )`property ${prop} was not expected`;
+  }
+  for (const prop of expectedProperties) {
+    insist(
+      actualProperties.includes(prop),
+    )`property ${prop} was expected but not present`;
+  }
+  return true;
+};
+
+// Installation Table
+// Columns: handle | installation
+const makeInstallationTable = () => {
+  const validate = obj => validateProperties(['installation'], obj);
+  return makeTable(validate);
+};
+
+// Instance Table
+// Columns: handle | installationHandle | instance | terms | assays
+const makeInstanceTable = () => {
+  const validate = obj =>
+    validateProperties(
+      ['installationHandle', 'instance', 'terms', 'assays'],
+      obj,
+    );
+  return makeTable(validate);
+};
+
+// Offer Table
+// Columns: handle | instanceHandle | assays | payoutRules | exitRule
+// | payoutPromise | units | extents
+const makeOfferTable = () => {
+  const insistValidPayoutRuleKinds = payoutRules => {
+    const acceptedKinds = [
+      'offerExactly',
+      'offerAtMost',
+      'wantExactly',
+      'wantAtLeast',
+    ];
+    for (const payoutRule of payoutRules) {
+      insist(
+        acceptedKinds.includes(payoutRule.kind),
+      )`${payoutRule.kind} must be one of the accepted kinds.`;
+    }
   };
+  const insistValidExitRule = exitRule => {
+    const acceptedExitRuleKinds = [
+      'noExit',
+      'onDemand',
+      'afterDeadline',
+      // 'onDemandAfterDeadline', // not yet supported
+    ];
+    insist(
+      acceptedExitRuleKinds.includes(exitRule.kind),
+    )`exitRule.kind ${exitRule.kind} is not one of the accepted options`;
+  };
+
+  const validate = obj => {
+    validateProperties(
+      [
+        'instanceHandle',
+        'assays',
+        'payoutRules',
+        'exitRule',
+        'payoutPromise',
+        'units',
+        'extents',
+      ],
+      obj,
+    );
+    insistValidPayoutRuleKinds(obj.payoutRules);
+    insistValidExitRule(obj.exitRule);
```

Comment:
These tests are not adequate to determine that you actually have a valid offer. Malformed offers will pass these tests. What do you intend here?

Reply:
Just recording what we talked over in person - I think we agreed that this PR isn't meant to protect against malformed offers, but that we should make a note of that. I changed the name of the `validate` functions to be `validateSomewhat` and added TODOs to cover protection against malformed offer